rust/tw_aptos/src/transaction_builder.rs

```diff
@@ -0,0 +1,225 @@
+// Copyright © 2017-2023 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+use std::str::FromStr;
+use move_core_types::account_address::AccountAddress;
+use tw_keypair::traits::KeyPairTrait;
+use tw_proto::Aptos::Proto::mod_SigningInput::OneOftransaction_payload;
+use crate::constants::{GAS_UNIT_PRICE, MAX_GAS_AMOUNT};
+use crate::transaction::RawTransaction;
+use crate::transaction_payload::{EntryFunction, TransactionPayload};
+use tw_proto::Aptos::Proto::SigningInput;
+use crate::aptos_move_packages::aptos_account_transfer;
+
+pub struct TransactionBuilder {
+    sender: Option<AccountAddress>,
+    sequence_number: Option<u64>,
+    payload: TransactionPayload,
+    max_gas_amount: u64,
+    gas_unit_price: u64,
+    expiration_timestamp_secs: u64,
+    chain_id: u8,
+}
+
+impl TransactionBuilder {
+    pub fn new(
+        payload: TransactionPayload,
+        expiration_timestamp_secs: u64,
+        chain_id: u8,
+    ) -> Self {
+        Self {
+            payload,
+            chain_id,
+            expiration_timestamp_secs,
+            max_gas_amount: MAX_GAS_AMOUNT,
+            gas_unit_price: std::cmp::max(GAS_UNIT_PRICE, 1),
+            sender: None,
+            sequence_number: None,
+        }
+    }
+
+    pub fn sender(mut self, sender: AccountAddress) -> Self {
+        self.sender = Some(sender);
+        self
+    }
+
+    pub fn sequence_number(mut self, sequence_number: u64) -> Self {
+        self.sequence_number = Some(sequence_number);
+        self
+    }
+
+    pub fn max_gas_amount(mut self, max_gas_amount: u64) -> Self {
+        self.max_gas_amount = max_gas_amount;
+        self
+    }
+
+    pub fn gas_unit_price(mut self, gas_unit_price: u64) -> Self {
+        self.gas_unit_price = gas_unit_price;
+        self
+    }
+
+    pub fn chain_id(mut self, chain_id: u8) -> Self {
+        self.chain_id = chain_id;
+        self
+    }
+
+    pub fn expiration_timestamp_secs(mut self, expiration_timestamp_secs: u64) -> Self {
+        self.expiration_timestamp_secs = expiration_timestamp_secs;
+        self
+    }
+
+    pub fn build(self) -> RawTransaction {
+        RawTransaction::new(
+            self.sender.expect("sender must have been set"),
+            self.sequence_number
+                .expect("sequence number must have been set"),
```

Comment:
I'm very skeptical about using `expect` and `unwrap` in builder pattern, because at this moment you are sure that `TransactionBuilder.sender()` and `TransactionBuilder.sequence_number()` methods were called, but some day someone can forget to call them, or the interface will be changed etc.
I prefer returning an error if some of the builder members not set, for example, `SigningError`.

Or as an alternative, I would rather construct `RawTransaction` in place instead of using the `TransactionBuilder`, because builder pattern is very useful when there is some logic in `build()` method, but in this case it just forwards the values to the `RawTransaction::new`.

Reply:
> I'm very skeptical about using `expect` and `unwrap` in builder pattern, because at this moment you are sure that `TransactionBuilder.sender()` and `TransactionBuilder.sequence_number()` methods were called, but some day someone can forget to call them, or the interface will be changed etc. I prefer returning an error if some of the builder members not set, for example, `SigningError`.
> 
> Or as an alternative, I would rather construct `RawTransaction` in place instead of using the `TransactionBuilder`, because builder pattern is very useful when there is some logic in `build()` method, but in this case it just forwards the values to the `RawTransaction::new`.

I'm trying to maintains the same approach as https://github.com/aptos-labs/aptos-core/blob/dcf24abbffa959e6aac5c832ca2f690441e32a97/sdk/src/transaction_builder.rs so it's easier to update in the future, wdyt?

Reply:
Makes sense 👍